R/generate_projections.R

```diff
@@ -0,0 +1,207 @@
+
+# real_time_projection() -> get_mspe()
+#
+#                    IS IN
+#
+# evaluate_real_time_projection()
+
+
+#### inverse Z score #####
+#' @description Inversely transform z-score standardized data to raw data
+#'              using: raw = z*sd + mean
+#'
+#' @param old_data: the previous data that is used for z-score standardization
+#' @param scaled_data: current scaled data
+#' @return inversely z-score transformed data
+inv_z_scale <- function(old_data, scaled_data) {
+    sd <- sd(old_data)
+    m <- mean(old_data)
+
+    scaled_data * sd + m
+}
+
+
+#' @description Data format preparation for model building.
+#' It includes:
+#' 1. Add `previous` predictor.
+#' 2. Z-score standardization of `previous`,`elapsed`, and `daily`
+#' @param data: the data frame need to be prepared.
+#' Must include columns:`previous`,`elapsed`, and `daily`
+#' @return Data frame with scaled `previous`,`elapsed`, and `daily` and ready to build model
+data_prepare <- function(data) {
+    data$previous <- c(NA, data$daily[-nrow(data)]) # nolint
+    data <- data[-1, ]
+    data$previous_std <- scale(data$previous, center = T, scale = T)
+    data$elapsed_std <- scale(data$elapsed, center = T, scale = T)
+    data$daily_std <- scale(data$daily, center = T, scale = T)
+
+    return(data)
+}
+
+#' @description This function uses `train_data` to train the linear regression
+#' and project to the last date in the `test_data`.
+#' It is embedded in `plot_real_time_projection` and 'evaluate_real_time_projection'
+#' @param train_data: data frame used to train linear regression
+#' @param test_data: data frame used to compare with the prediction generated by linear regression
+#' @return a list containing the predicted uptake and the corresponding
+#' observed rate under the `test_data` time frame,
+#' in both uptake rate `rate` and cumulative uptake `cumulative`
+
+real_time_projection <- function(
+    train_data = nis_usa_2022,
+    test_data = nis_usa_2023) {
+    # transform the train data and test data to the format of model building and model prediction #
+    scaled_train <- data_prepare(train_data)
+    scaled_test <- data_prepare(test_data)
+
+    # The first row is also removed for raw test data, to be comparable with the predicted values #
+    test_data <- test_data[-1, ]
+
+    # model building #
+    model <- build_projection_model2(data = scaled_train)
+
+    ## get out-of-sample prediction ##
+    scaled_pred <- brms::posterior_predict(model, newdata = scaled_test)
+
+    ## recover the prediction to the original format for comparision ##
+    pred <- inv_z_scale(test_data$daily, scaled_pred)
+
+    # predicted daily vaccine uptake #
+    pred_summary <- data.frame(
+        date = test_data$date,
+        obs = test_data$daily,
+        lower = apply(pred, 2, quantile, 0.025),
+        upper = apply(pred, 2, quantile, 0.975),
+        mean = colMeans(pred)
+    )
+
+    # predicted cumulative vaccine uptake #
+    cumu <- t(apply(pred, 1, cumsum))
```

Comment:
Cumulative sums of incident uptake alone will be too small, because we also need to include the cumulative uptake from the first 1 or 2 data points of the season, which were dropped prior to model fitting.